### ISSUE TITLE
Add tensor typing

### ISSUE THREAD

**Issue by juanmc2005** [juanmc2005/StreamingSpeakerDiarization#63]
## Problem

It's difficult to understand the shape and format expected from tensors.

## Idea

Add [torchtyping](https://github.com/patrick-kidger/torchtyping).
There's also the possibility to check that types are respected at runtime (instead of it being for documentation only), for example while debugging.

**Comment by manish-kumar-iisc** [comment]
@juanmc2005 Can i work on this ?

**Comment by juanmc2005** [comment]
@manish-kumar-iisc absolutely. This dates from a while back though so there may be better options out there now.
What did you have in mind?

**Comment by manish-kumar-iisc** [comment]
I will implement tensor typing support using torchtyping + typeguard with a focus on optional, zero-overhead integration. I will try to do:

Implementation Overview
1. New tensor_typing.py helper module — Provides:
	a) TensorType for shape annotations (falls back to typing.Any if torchtyping isn't installed)
	b) get_typechecked_decorator() for optional runtime shape validation
	c) Only activates when DIART_DEBUG_TENSOR_CHECKS=1 is set — no performance impact in production

2. Annotated core public APIs in models.py:
	a) PowersetAdapter.forward() — Input/output shapes documented
	b) SegmentationModel.__call__() — Added @typechecked decorator
	c) EmbeddingModel.__call__() — Handles optional weights parameter

3. Testing & Documentation:
	a) Unit test validating runtime shape checks work correctly
	b) requirements-dev.txt with optional dev dependencies
	c) Documentation explaining how to enable/disable checks

Happy to revise based on your preference — which direction should I take? Let me know your feedback.

**Comment by juanmc2005** [comment]
That sounds about right! I'd really like to move to a `pyproject.toml` now, I opened PR #286 for this, I just still need to fix the tests and CI. Maybe you could branch from there? I think that will make your life way easier for this implementation
